unix/w0vncserver/w0vncserver.man
.TH w0vncserver 1 "" "TigerVNC" "TigerVNC Manual"
.SH NAME
w0vncserver \- TigerVNC server for Wayland compositors
.SH SYNOPSIS
.B w0vncserver
.RI [ options ]
.br
.B w0vncserver -version
.SH DESCRIPTION
.B w0vncserver
is a TigerVNC server which makes a Wayland compositor that is based on
wlroots, or has the RemoteDesktop portal implemented, remotely
accessible via VNC, TigerVNC or compatible viewers. It does not create
a virtual display, instead, it shares an existing display (typically,
hat one connected to the physical screen).

.SH OPTIONS
.B w0vncserver
interprets the command line as a list of parameters with optional values.
Running \fBw0vncserver \-h\fP will show a list of all valid parameters with
short descriptions.  All parameters are optional, but normally you would have
to use the \fBPasswordFile\fP parameter (see its description below).
.PP
There are several forms of specifying parameters in the command line (here we
use `\fISomeParameter\fP' as an example parameter name):

.TP
.B -\fISomeParameter\fP
Enable the parameter, turn the feature on.  This form can be used with
parameters that simply enable or disable some feature.
.
.TP
.B -\fISomeParameter\fP=0
Disable the parameter, turn the feature off.
.
.TP
.B -\fISomeParameter\fP=\fIvalue\fP
Assign the specified \fIvalue\fP to the parameter.  The leading dash can be
omitted, or it can be doubled if desired (like in GNU-style long options).
.PP
Parameter names are case-insensitive, their order in the command line can be
arbitrary.

.SH PARAMETERS

.TP
.B \-AcceptCutText
Accept clipboard updates from clients. Default is on.
.
.TP
.B \-AcceptKeyEvents
Accept key press and release events from clients. Default is on.
.
.TP
.B \-AcceptPointerEvents
Accept pointer movement and button events from clients. Default is on.
.
.TP
.B \-AlwaysShared
Always treat incoming connections as shared, regardless of the client-specified
setting. Default is off.
.
.TP
.B \-BlacklistThreshold \fIcount\fP
The number of unauthenticated connection attempts allowed from any individual
host before that host is black-listed.  Default is 5.
.
.TP
.B \-BlacklistTimeout \fIseconds\fP
The initial timeout applied when a host is first black-listed.  The host
cannot re-attempt a connection until the timeout expires.  Default is 10.
.
.TP
.B \-CompareFB \fImode\fP
Perform pixel comparison on framebuffer to reduce unnecessary updates. Can
be either \fB0\fP (off), \fB1\fP (always) or \fB2\fP (auto). Default is
\fB2\fP.
.
.TP
.B \-desktop \fIdesktop-name\fP
Each desktop has a name which may be displayed by the viewer. It defaults to
"<user>@<hostname>".
.
.TP
.B \-DisconnectClients
Disconnect existing clients if an incoming connection is non-shared. Default is
on. If \fBDisconnectClients\fP is false, then a new non-shared connection will
be refused while there is a client active.  When combined with
\fBNeverShared\fP this means only one client is allowed at a time.
.
.TP
.B \-FrameRate \fIfps\fP
The maximum number of updates per second sent to each client. If the screen
updates any faster then those changes will be aggregated and sent in a single
update to the client. Note that this only controls the maximum rate and a
client may get a lower rate when resources are limited. Default is \fB60\fP.
.
.TP
.B \-GnuTLSPriority \fIpriority\fP
GnuTLS priority string that controls the TLS session’s handshake algorithms.
See the GnuTLS manual for possible values. Default is \fBNORMAL\fP.
.
.TP
.B \-IdleTimeout \fIseconds\fP
The number of seconds after which an idle VNC connection will be dropped.
Default is 0, which means that idle connections will never be dropped.
.
.TP
.B \-ImprovedHextile
Use improved compression algorithm for Hextile encoding which achieves better
compression ratios by the cost of using slightly more CPU time.  Default is
on.
.
.TP
.B \-interface \fIIP address\fP
Listen on interface. By default w0vncserver listens on all available interfaces.
.
.TP
.B \-localhost
Only allow connections from the same machine. Useful if you use SSH and want to
stop non-SSH connections from any other hosts.
.
.TP
.B \-dnssd
Advertise server via mDNS/DNS-SD. Default is on.
.
.TP
.B \-Log \fIlogname\fP:\fIdest\fP:\fIlevel\fP[, ...]
Configures the debug log settings.  \fIdest\fP can currently be \fBstderr\fP,
\fBstdout\fP or \fBsyslog\fP, and \fIlevel\fP is between 0 and 100, 100 meaning
most verbose output.  \fIlogname\fP is usually \fB*\fP meaning all, but you can
target a specific source file if you know the name of its "LogWriter".  Default
is \fB*:stderr:30\fP.
.
.TP
.B \-MaxConnectionTime \fIseconds\fP
Terminate when a client has been connected for \fIN\fP seconds.  Default is
0.
.TP
.B \-MaxCutText \fIbytes\fP
The maximum permitted size of an incoming clipboard update.
Default is \fB262144\fP.
.
.
.TP
.B \-MaxDisconnectionTime \fIseconds\fP
Terminate when no client has been connected for \fIN\fP seconds.  Default is
0.
.
.TP
.B \-MaxIdleTime \fIseconds\fP
Terminate after \fIN\fP seconds of user inactivity.  Default is 0.
.
.TP
.B \-NeverShared
Never treat incoming connections as shared, regardless of the client-specified
setting. Default is off.
.
.TP
.B \-pam_service \fIname\fP, \-PAMService \fIname\fP
PAM service name to use when authentication users using any of the "Plain"
security types. Default is \fBvnc\fP.
.
.TP
.B \-Password \fIpassword\fP
Obfuscated binary encoding of the password which clients must supply to
access the server.  Using this parameter is insecure, use \fBPasswordFile\fP
parameter instead.
.
.TP
.B \-PasswordFile \fIpasswd-file\fP, \-rfbauth \fIpasswd-file\fP
Password file for VNC authentication.  There is no default, you should
specify the password file explicitly.  Password file should be created with
the \fBvncpasswd\fP(1) utility.  The file is accessed each time a connection
comes in, so it can be changed on the fly.
.
.TP
.B \-PlainUsers \fIuser-list\fP
A comma separated list of user names that are allowed to authenticate via
any of the "Plain" security types (Plain, TLSPlain, etc.). Specify \fB*\fP
to allow any user to authenticate using this security type. Specify \fB%u\fP
to allow the user of the server process. Default is to deny all users.
.
.TP
.B \-Protocol3.3
Always use protocol version 3.3 for backwards compatibility with badly-behaved
clients. Default is off.
.
.TP
.B \-QueryConnect
Prompts the user of the desktop to explicitly accept or reject incoming
connections. Default is off.
.
.TP
.B \-RawKeyboard
Send keyboard events straight through and avoid mapping them to the current
keyboard layout. This effectively makes the keyboard behave according to the
layout configured on the server instead of the layout configured on the
client. Default is off.
.
.TP
.B \-RemapKeys \fImapping
Sets up a keyboard mapping.
.I mapping
is a comma-separated string of character mappings, each of the form
.IR char -> char ,
or
.IR char <> char ,
where
.I char
is a hexadecimal keysym. For example, to exchange the " and @ symbols you would specify the following:

.RS 10
RemapKeys=0x22<>0x40
.RE
.
.TP
.B \-RememberDisplayChoice \fImode\fP
Remember display choice when a user connects. Can be either
\fBAlways\fP, \fBNever\fP, or \fBSession\fP. Choosing \fBNever\fP means
a prompt dialog will be be shown every time a user connects to the
server. \fBSession\fP means a prompt will be shown only once during the
program's lifetime. When choosing \fBAlways\fP, the choice you make will
be remembered and persist. To reset your choice, see
\fBw0vncserver-forget\fP(1). Default is \fBSession\fP.
.
.TP
.B \-RequireUsername
Require username for the RSA-AES security types. Default is off.
.
.TP
.B \-rfbport \fIport\fP
Specifies the TCP port on which w0vncserver listens for connections from
viewers (the protocol used in VNC is called RFB - "remote framebuffer").
Specify \fB-1\fP to disable listening on a TCP port. The default port is
5900 when started directly, and -1 when activated by a systemd socket.
.
.TP
.B \-rfbunixmode \fImode\fP
Specifies the mode of the Unix domain socket.  The default is 0600.
.
.TP
.B \-rfbunixpath \fIpath\fP
Specifies the path of a Unix domain socket on which w0vncserver listens for
connections from viewers. Default is to not listen to any Unix domain
socket.
.
.TP
.B \-RSAKey \fIpath\fP
Path to the RSA key for the RSA-AES security types (\fBRA2\fP, \fBRA2ne\fP,
\fBRA2_256\fP and \fBRA2ne_256\fP) in PEM format.
.
.TP
.B \-SecurityTypes \fIsec-types\fP
Specify which security scheme to use for incoming connections.  Valid values
are a comma separated list of \fBNone\fP, \fBVncAuth\fP, \fBPlain\fP,
\fBTLSNone\fP, \fBTLSVnc\fP, \fBTLSPlain\fP, \fBX509None\fP, \fBX509Vnc\fP,
\fBX509Plain\fP, \fBRA2\fP, \fBRA2ne\fP, \fBRA2_256\fP and \fBRA2ne_256\fP.
Default is \fBTLSVnc,VncAuth\fP.
.
.TP
.B \-SendCutText
Send clipboard changes to clients. Default is on.
.
.B \-SendPrimary
Send the primary as well as the selection clipboard to clients. Note
that this does not work on all compositors. Default is on.
.
.TP
.B \-SetPrimary
Set the primary as well as the selection clipboard. Note that this does
not work on all compositors. Default is on.
.
.TP
.B \-UseBlacklist
Temporarily reject connections from a host if it repeatedly fails to
authenticate. Default is on.
.
.TP
.B \-UseIPv4
Use IPv4 for incoming and outgoing connections. Default is on.
.
.TP
.B \-UseIPv6
Use IPv6 for incoming and outgoing connections. Default is on.
.
.TP
.B \-X509Cert \fIpath\fP
Path to a X509 certificate in PEM format to be used for all X509 based
security types (X509None, X509Vnc, etc.).
.
.TP
.B \-X509Key \fIpath\fP
Private key counter part to the certificate given in \fBX509Cert\fP. Must
also be in PEM format.

.SH SEE ALSO
.BR w0vncserver-forget (1),
.BR x0vncserver (1),
.BR Xvnc (1),
.BR vncpasswd (1),
.br
https://www.tigervnc.org/

.SH AUTHOR
TigerVNC Team

VNC was originally developed by the RealVNC team while at Olivetti
Research Ltd / AT&T Laboratories Cambridge.  TightVNC additions were
implemented by Constantin Kaplinsky. Many other people have since
participated in development, testing and support. This manual is part
of the TigerVNC software suite.


unix/w0vncserver/w0vncserver.cxx
/* Copyright 2025 Adam Halim for Cendio AB
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <assert.h>

#include <glib.h>
#include <glib-unix.h>
#include <gio/gio.h>

#include <network/TcpSocket.h>
#include <network/UnixSocket.h>
#include <rfb/VNCServerST.h>
#include <rfb/ServerCore.h>
#include <rdr/FdOutStream.h>
#include <core/Logger_stdio.h>
#include <core/LogWriter.h>
#include <core/i18n.h>

#include "parameters.h"
#include "GSocketSource.h"
#include "RFBTimerSource.h"
#include "portals/PortalDesktop.h"
#include "wayland/WaylandDesktop.h"

static core::LogWriter vlog("main");

char* programName;
static GMainLoop* loop = nullptr;
static GDBusConnection* connection = nullptr;
static bool fatalError = false;

void fatal_error(const char *error, ...) {
  char exitError[1024];

  // Prioritise the first error we get as that is probably the most
  // relevant one.
  if (fatalError)
    return;

  fatalError = true;

  va_list ap;
  va_start(ap, error);
  vsnprintf(exitError, sizeof(exitError), error, ap);
  va_end(ap);

  vlog.error("%s", exitError);

  assert(loop);
  g_main_loop_quit(loop);
}

static void printVersion(FILE *fp)
{
  fprintf(fp, _("TigerVNC server version %s, built %s\n"),
          PACKAGE_VERSION, BUILD_TIMESTAMP);
}

static void usage()
{
  printVersion(stderr);
  fprintf(stderr,
          _("\n"
            "Usage: %s [<parameters>]\n"
            "       %s --version\n"),
          programName, programName);
  fprintf(stderr,
          _("\n"
            "Parameters can be turned on with -<param> or off with -<param>=0\n"
            "Parameters which take a value can be specified as "
            "-<param> <value>\n"
            "Other valid forms are <param>=<value> -<param>=<value> "
            "--<param>=<value>\n"
            "Parameter names are case-insensitive.  The parameters are:\n\n"));
  core::Configuration::listParams(79, 14);
  exit(1);
}

int CleanupSignalHandler(void*)
{
  g_main_loop_quit(loop);
  return 1;
}

int main(int argc, char** argv)
{
  std::list<network::SocketListener*> listeners;
  int32_t sigintTag;
  int32_t sigtermTag;
  int32_t sighupTag;
  RFBTimerSource* timerSource;
  GSocketSource* monitor;
  rfb::SDesktop* desktop;
  rfb::VNCServerST* server;

  setlocale(LC_ALL, "");

  programName = argv[0];

  core::initStdIOLoggers();
  core::LogWriter::setLogParams("*:stderr:30");

  core::Configuration::removeParam("AcceptSetDesktopSize");

  for (int i = 1; i < argc;) {
    int ret;

    ret = core::Configuration::handleParamArg(argc, argv, i);
    if (ret > 0) {
      i += ret;
      continue;
    }

    if (strcmp(argv[i], "-h") == 0 ||
        strcmp(argv[i], "-help") == 0 ||
        strcmp(argv[i], "--help") == 0) {
      usage();
    }

    if (strcmp(argv[i], "-v") == 0 ||
        strcmp(argv[i], "-version") == 0 ||
        strcmp(argv[i], "--version") == 0) {
      printVersion(stderr);
      return 0;
    }

    if (argv[i][0] == '-') {
      fprintf(stderr, _("%s: Unrecognized option '%s'\n"),
              programName, argv[i]);
      fprintf(stderr, _("See '%s --help' for more information.\n"),
              programName);
      exit(1);
    }

    fprintf(stderr, _("%s: Extra argument '%s'\n"), programName, argv[i]);
    fprintf(stderr, _("See '%s --help' for more information.\n"),
            programName);
    exit(1);
  }

  try {
    if (rfbunixpath.getValueStr()[0] != '\0') {
      listeners.push_back(new network::UnixListener(rfbunixpath, rfbunixmode));
      vlog.info(_("Listening for VNC connections on %s (mode %04o)"),
                (const char*)rfbunixpath, (int)rfbunixmode);
    }

    if ((int)rfbport != -1) {
      std::list<network::SocketListener*> tcp_listeners;
      const char *addr = interface;

      if (strcasecmp(addr, "all") == 0)
        addr = nullptr;
      if (localhostOnly)
        createLocalTcpListeners(&tcp_listeners, (int)rfbport);
      else
        createTcpListeners(&tcp_listeners, addr, (int)rfbport);

      if (!tcp_listeners.empty()) {
        listeners.splice (listeners.end(), tcp_listeners);
        if (localhostOnly)
          vlog.info(_("Listening for VNC connections on local "
                      "interfaces, port %d"), (int)rfbport);
        else
          vlog.info(_("Listening for VNC connections on interface %s, "
                      "port %d"),
                    (const char*)interface, (int)rfbport);
      }
    }
  } catch (std::exception& e) {
    fprintf(stderr, "%s\n", e.what());
    return -1;
  }

  if (listeners.empty()) {
    vlog.error(_("No path or port configured for incoming "
                  "connections"));
    return -1;
  }

  loop = g_main_loop_new(nullptr, false);
  sigintTag = g_unix_signal_add(SIGINT, CleanupSignalHandler, nullptr);
  sigtermTag = g_unix_signal_add(SIGTERM, CleanupSignalHandler, nullptr);
  sighupTag = g_unix_signal_add(SIGHUP, CleanupSignalHandler, nullptr);

  try {
    if (PortalDesktop::available()) {
      desktop = new PortalDesktop();
    } else if (WaylandDesktop::available()) {
      desktop = new WaylandDesktop(loop);
    } else {
      vlog.error(_("Desktop does not support remote connections"));
      return -1;
    }
    server = new rfb::VNCServerST(desktopName, desktop);
    timerSource = new RFBTimerSource();
    monitor = new GSocketSource(server, &listeners);
  } catch (std::exception& e) {
    vlog.error("main: %s", e.what());
    return 1;
  }

  monitor->attach(g_main_loop_get_context(loop));
  timerSource->attach(g_main_loop_get_context(loop));
  monitor->listen();
  g_main_loop_run(loop);

  g_source_remove(sigintTag);
  g_source_remove(sigtermTag);
  g_source_remove(sighupTag);

  delete monitor;
  delete timerSource;
  delete server;
  delete desktop;

  for (network::SocketListener* listener : listeners)
    delete listener;

  // The GDBusConnection is a singleton that we want to close once
  // before we exit.
  connection = g_bus_get_sync(G_BUS_TYPE_SESSION, nullptr, nullptr);
  if (connection)
    g_dbus_connection_close_sync(connection, nullptr, nullptr);

  g_main_loop_unref(loop);
}
